Nodejs/Product/Nodejs/SharedProject/CommonProjectNode.cs

```diff
@@ -399,27 +403,30 @@ protected override void Reload()
             this.watcher = CreateFileSystemWatcher(this.ProjectHome);
             this.attributesWatcher = CreateAttributesWatcher(this.ProjectHome);
 
-            this.currentMerger = new DiskMerger(this, this, this.ProjectHome);
+            this.diskMerger.Clear();
+            this.diskMerger.AddDir(this, this.ProjectHome);
+            this.diskMerger.Start();
         }
 
         /// <summary>
         /// Called to ensure that the hierarchy's show all files nodes are in
         /// sync with the file system.
         /// </summary>
-        protected void SyncFileSystem()
+        protected void WaitForSyncFileSystem()
         {
-            if (this.currentMerger == null)
-            {
-                this.currentMerger = new DiskMerger(this, this, this.ProjectHome);
-            }
-            while (this.currentMerger.ContinueMerge(this.ParentHierarchy != null))
-            {
-            }
-            this.currentMerger = null;
+            // ensure the merge is started, and wait for completion
+            // this blocks the UI thread
```

Comment:
Doesn't this just block whatever thread it is called on (which may or may not be the UI thread).

Also, does that risk deadlocking? If this is blocking the UI thread, but there are many places in `Merge` where you call `InvokeOnUIThread`.